### ISSUE TITLE
Limitations of Routing Slips in Wolverine Sagas

### ISSUE THREAD

**Issue by PascalSenn** [JasperFx/wolverine#237]
Wolverine implements sagas by using routing slips, as outlined here (link form the docs): https://www.enterpriseintegrationpatterns.com/patterns/messaging/RoutingTable.html

However, for me, a saga is more than just a routing slip, as usually a saga also includes fault compensation, which is not covered by the current implementation

This paper (https://www.cs.cornell.edu/andru/cs711/2002fa/reading/sagas.pdf) explains this in detail, but this blog post is easier to read: https://medium.com/design-microservices-architecture-with-patterns/saga-pattern-for-microservices-distributed-transactions-7e95d0613345

The idea behind fault compensation is that an orchestrator keeps a log of what has been executed and if an error occurs, the orchestrator can call predefined compensating transactions to undo the previous steps.

For example, if we have three transactions: Create User, Create Address, and Ship Product, and there's an error in the last step, the orchestrator will call the compensating transactions in reverse order: Delete Address, and Delete User.

```
Transaction1: Create User  => LOG: Created User with id 1
Transaction2: Create Address => LOG: Created Address for user with id 2
Transaction3: Ship Product => Error: Product not found
Compensate3:
Compensate2: Delete Address with Id 2
Compensate1: Delete User with Id 1
```

**Comment by PascalSenn** [comment]
Oops, my bad. I was going through the documentation of Masstransit and Wolverine late yesterday and must have messed up the links. I thought the [Routing Table](https://www.enterpriseintegrationpatterns.com/patterns/messaging/RoutingTable.html) was referenced in the docs, but it was actually the [Process Manager](https://www.enterpriseintegrationpatterns.com/patterns/messaging/ProcessManager.html). In my defense, MassTransit references the routing slip of EIP in their docs, and the pages look very similar.

Here the part of the docs:
> As is so common in these docs, I would direct you to this from the old "EIP" book: [Process Manager](http://www.enterpriseintegrationpatterns.com/patterns/messaging/ProcessManager.html).


Regarding:
> It's a great set of links, but what are you looking for here?

It's a bit of a terminology thing I had trouble understanding. Masstransit does the same thing; their "Saga" is not really a "Saga" is it? Or what's the definition of a Saga that is used here?

Regarding: 
> What do you need right now?

I don't need anything right now. I may be interested in sagas down the road. I just wanted to place this here and may even pick it up later as a contribution. It's nice to have an orchestrator, log, and compensating transactions when you have complex workflows to implement, which may even be distributed across services.

No immediate action to be taken;I guess i just want to know what you think about this. This issue can also be converted to a Discussion.

**Comment by jeremydmiller** [comment]
"MassTransit references the routing slip of EIP in their docs, and the pages look very similar." -- everybody is using VitePress/VuePress, so all the doc sites look *somewhat* similar. I thought Chris had just moved the docs to something new, and I haven't even checked it out yet.

"I guess i just want to know what you think about this." -- honestly, I think it's complicated and it was something that I was purposely putting off until folks asked for it or I had a concrete need for it at work. It's not something I want to build outside of the context of a real world example. I think a stateful saga would get it done for multi-step workflows, and you could author the compensating actions manually. 

I'm not ruling out building a routing slip implementation later, and I'd be very happy for contributions in that regard. Especially if there's a real use case.

It's getting close to the point where Wolverine has to be all doc updates & bug fixes to get to 1.0 very soon because that's lingering too long.

**Comment by PascalSenn** [comment]
Alirghty let's close this issue then so someone can pick it up when the need arises for it.
